internal/repository/index.go

```diff
@@ -214,35 +187,30 @@ func (idx *Index) Lookup(id restic.ID, tpe restic.BlobType) (blobs []restic.Pack
 	idx.m.Lock()
 	defer idx.m.Unlock()
 
-	h := restic.BlobHandle{ID: id, Type: tpe}
-
-	blob, ok := idx.blob[h]
-	if ok {
-		blobList := idx.withDuplicates(h, blob)
-		blobs = make([]restic.PackedBlob, 0, len(blobList))
+	idx.mapFor(tpe).foreachWithID(id, func(e *indexEntry) {
+		blobs = append(blobs, idx.toPackedBlob(e, tpe))
+	})
 
-		for _, p := range blobList {
-			blobs = append(blobs, idx.indexEntryToPackedBlob(h, p))
-		}
-
-		return blobs, true
-	}
-
-	return nil, false
+	return blobs, len(blobs) > 0
 }
 
 // ListPack returns a list of blobs contained in a pack.
 func (idx *Index) ListPack(id restic.ID) (list []restic.PackedBlob) {
 	idx.m.Lock()
 	defer idx.m.Unlock()
 
-	for h, entry := range idx.blob {
-		for _, blob := range idx.withDuplicates(h, entry) {
-			if idx.packs[blob.packIndex] == id {
-				list = append(list, idx.indexEntryToPackedBlob(h, blob))
-			}
+	idx.data.foreach(func(e *indexEntry) bool {
+		if idx.packs[e.packIndex] == id {
+			list = append(list, idx.toPackedBlob(e, restic.DataBlob))
 		}
-	}
+		return true
```

Comment:
This makes `foreach` return after the first blob in this pack. It should loop over all entries and append all blobs with matching pack ID.

Reply:
No, the callback returns true to signal that it wants to continue. When this is set to false, TestIndexUnserialize fails.

Reply:
Ah - I missed the `!` in `foreach`. Then it's ok!